chromadb/db/__init__.py

```diff
@@ -1,125 +1,155 @@
 from abc import ABC, abstractmethod
-from typing import Dict, List, Sequence, Optional, Tuple
+from typing import Optional, TypedDict, Union, Literal
 from uuid import UUID
+from collections.abc import Sequence
 import numpy.typing as npt
-from chromadb.api.types import Embeddings, Documents, IDs, Metadatas, Where, WhereDocument
 
 
-class DB(ABC):
-    @abstractmethod
-    def __init__(self):
-        pass
+StrDict = Optional[dict[str, str]]
+NameOrID = Union[str, UUID]
 
-    @abstractmethod
-    def create_collection(
-        self, name: str, metadata: Optional[Dict] = None, get_or_create: bool = False
-    ) -> Sequence:
-        pass
 
-    @abstractmethod
-    def get_collection(self, name: str) -> Sequence:
-        pass
+class Collection(TypedDict):
+    id: UUID
+    name: str
+    metadata: StrDict = None
 
-    @abstractmethod
-    def list_collections(self) -> Sequence:
-        pass
 
-    @abstractmethod
-    def update_collection(
-        self, current_name: str, new_name: Optional[str] = None, new_metadata: Optional[Dict] = None
-    ):
-        pass
+class Segment(TypedDict):
+    id: UUID
+    python_class: str
+    metadata: StrDict
 
-    @abstractmethod
-    def delete_collection(self, name: str):
-        pass
 
-    @abstractmethod
-    def get_collection_uuid_from_name(self, collection_name: str) -> str:
-        pass
+class System(ABC):
+    """Interface for Chroma's System storage backend"""
+
 
     @abstractmethod
-    def add(
-        self,
-        collection_uuid: str,
-        embeddings: Embeddings,
-        metadatas: Optional[Metadatas],
-        documents: Optional[Documents],
-        ids: List[UUID],
-    ) -> List[UUID]:
-        pass
+    def create_collection(self,
```

Comment:
Is a collection created with one segment by default?

Reply:
For the embedded and single-server use cases this week, yes, my plan was to create a single segment for each collection on demand (in the `get_segment` method). This will actually work for simple partitioning as well.

For the distributed case, we still need more thought on how/when to create segments since we'll probably want to utilize the same mechanics that we use for rolling over/partitioning etc. My current thought is a "partitioning strategy" entity stored in the SysDB that can be associated with a collection.

Reply:
Ok so it will be delegated. Cool.